### PR TITLE
[AWS] Add LB Type in the infrastructure cluster object via install-config yaml

### PR DESCRIPTION
**Pull request by gcs278**
**Note**: PR cloned from https://github.com/openshift/installer/pull/6074 so I could modify. This is @miheer's commit.

In the ingress config cluster object we need to add api for AwsLBType which will store and persist the information of the desired LB type mentioned by the installer config API so that the ingress operator will refer the LB type from this object whenever someone accidentally deletes the ingress controller.

Currently, the default LB type i.e CLB/ELB gets set if you delete the ingress controller object even when the desired the LB type for default ingress controller
was set to NLB as ingress operator does not have a way to check the LB type as we don't store it.
This commit adds LB Type for AWS mentioned in the install-config.yaml in the infrastructure cluster object.

`data/data/install.openshift.io_installconfigs.yaml` A generated new field called lbType will be available to add load balancer type in the install-config.yaml for AWS
`pkg/asset/manifests/ingress.go` The ingress cluster object will be patched with this lbType provided by user so that we have the information related to the lbType persistent which can be referred by ingress operator when the default ingress controller is deleted.
`pkg/asset/manifests/ingress_test.go` Unit test to test setting of lbType in the ingress cluster object's spec and status as per the value specified in the install-config.yaml
`pkg/explain/printer_test.go` A new field to test called lbType which will be available to add the load balancer type in the install-config.yaml for AWS
`pkg/types/aws/platform.go` A new field called lbType will be available to add the load balancer type in the install-config.yaml for AWS

Enhacement proposal - https://github.com/openshift/enhancements/pull/1148
OpenShift API - https://github.com/openshift/api/pull/1209

Jira ticket - https://issues.redhat.com/browse/NE-942

### CODE REVIEW

**Comment by gcs278** [comment]
I just removed the vendoring and openshift/api bump as the openshift/api version in master (`20221004120407-c46852673d03`) has our required changes for LB Type.

File: pkg/asset/manifests/ingress.go [hunk 1]
```diff
@@ -2,6 +2,7 @@ package manifests
 
 import (
 	"fmt"
+	"github.com/openshift/installer/pkg/types/aws"
```
Comment (r4f4) [inline pkg/asset/manifests/ingress.go:5]: Nit: this line should be in the installer imports block.

File: pkg/types/aws/platform.go [hunk 2]
```diff
@@ -59,6 +62,20 @@ type Platform struct {
 	// AWS resources that the operators create.
 	// +optional
 	PropagateUserTag bool `json:"propagateUserTags,omitempty"`
+
+	// LBType allows user to set a load balancer type.
+	// When this field is set the default ingresscontroller will get created using the specified LBType.
+	// If this field is not set then the default ingress controller of LBType Classic will be created.
```
[the remark this answers is not in the document]
Comment (gcs278) [reply pkg/types/aws/platform.go:68]: Yup - it's documented here [Shall we use NLB for default ingress controller ?](https://docs.google.com/document/d/1UmGog8S74h2SUHzCDbrkwbs0OqneTDLKRJsejPEWtXA/edit)
Specifically, we have this BZ https://bugzilla.redhat.com/show_bug.cgi?id=2023681 to figure out before going default.

File: pkg/asset/manifests/ingress.go (hunk 1 above)
Comment (gcs278) [reply pkg/asset/manifests/ingress.go]: Done

File: data/data/install.openshift.io_installconfigs.yaml [hunk 3]
```diff
@@ -1718,6 +1718,19 @@ spec:
...
+                      created using the specified LBType. If this field is not set
+                      then the default ingress controller of LBType Classic will be
+                      created. Valid values are: * "Classic": A Classic Load Balancer
+                      that makes routing decisions at either   the transport layer
+                      (TCP/SSL) or the application layer (HTTP/HTTPS). See   the following
+                      for additional details:   https://docs.aws.amazon.com/AmazonECS/latest/developerguide/load-balancer-types.html#clb
+                      * "NLB": A Network Load Balancer that makes routing decisions
+                      at the   transport layer (TCP/SSL). See the following for additional
+                      details:   https://docs.aws.amazon.com/AmazonECS/latest/developerguide/load-balancer-types.html#nlb'
+                    type: string
```
Comment (sadasu) [inline data/data/install.openshift.io_installconfigs.yaml:1733]: Nit: Some double spaces in the description fileld.

Comment (gcs278) [reply data/data/install.openshift.io_installconfigs.yaml:1733]: Hmm I see that, but I think this is generated from https://github.com/openshift/installer/pull/6478/files#diff-7941d8da46384fc43b6eeae7d5c3319065b16940457822346c9ccb08cbcf47b4R66, which uses a hanging indent and a bulleted list.

Should I remove the hanging indent and extra spaces from there?

Comment (r4f4) [reply data/data/install.openshift.io_installconfigs.yaml:1733]: I propose we do that as a follow-up.

**Comment by gcs278** [comment]
@sadasu Just modified the commit message to include my name in `Modified-by` for tracking purposes.

**Comment by miheer** [comment]
@r4f4  @sadasu @jhixson74 @mtulio  can we get this merged ASAP. We want this to get merged by today.

**Comment by Miciah** [comment]
I suspect this PR broke something.  I'm seeing `ingress.config.openshift.io "cluster" not found` errors in a lot of CI jobs including CI runs for merge pools that included this PR:

- https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/batch/pull-ci-openshift-installer-master-e2e-aws-ovn/1581059363588542464
- https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/batch/pull-ci-openshift-installer-master-e2e-aws-ovn/1580992330503032832
- https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/batch/pull-ci-openshift-installer-master-e2e-aws-ovn/1581019759770603520

I also see the same error in CI jobs for other repositories since this PR merged:

- https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_cluster-ingress-operator/843/pull-ci-openshift-cluster-ingress-operator-master-e2e-aws-operator/1581680873051262976
- https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_cluster-image-registry-operator/803/pull-ci-openshift-cluster-image-registry-operator-master-e2e-upgrade/1581264740628303872
- https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_cluster-control-plane-machine-set-operator/117/pull-ci-openshift-cluster-control-plane-machine-set-operator-main-e2e-aws-upgrade/1581652752310734848

See [search.ci](https://search.ci.openshift.org/?search=ingress.config.openshift.io+%22cluster%22+not+found&maxAge=168h&context=1&type=build-log&name=&excludeName=&maxMatches=5&maxBytes=20971520&groupBy=job) for more.

**Comment by Miciah** [comment]
In the `cluster-ingress-02-config.yml` manifest in the `artifacts/e2e-aws-ovn/ipi-install-install/artifacts/log-bundle-20221014201405.tar` artifact from <https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/batch/pull-ci-openshift-installer-master-e2e-aws-ovn/1580992330503032832>, I see the following:
```yaml
  loadBalancer:
    platform:
      aws: {}
      type: AWS
```
Whereas `loadBalancer`, `platform`, and `aws` are optional fields, the `type` subfield of the `aws` field is required:
https://github.com/openshift/api/blob/622889ac07cf3dd8787679af0d3176bf292b3185/config/v1/types_ingress.go#L144
Could that be causing the installer to fail to create the `ingress.config.openshift.io/cluster` object?

**Comment by Miciah** [comment]
Yup, `cluster-bootstrap-e0554618cff8.log` from the same artifacts shows the following:
```
"cluster-ingress-02-config.yml": failed to create ingresses.v1.config.openshift.io/cluster -n : Ingress.config.openshift.io "cluster" is invalid: spec.loadBalancer.platform.aws.type: Required value
```

**Pull request merged**